config/seata-config-core/src/main/java/io/seata/config/ConfigurationCache.java

```diff
@@ -0,0 +1,116 @@
+/*
+ *  Copyright 1999-2019 Seata.io Group.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.config;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import io.seata.common.util.CollectionUtils;
+import io.seata.common.util.StringUtils;
+import net.sf.cglib.proxy.Enhancer;
+import net.sf.cglib.proxy.MethodInterceptor;
+
+/**
+ * @author funkye
+ */
+public class ConfigurationCache implements ConfigurationChangeListener {
+
+    private static final String METHOD_PREFIX = "get";
+
+    private static final String METHOD_LATEST_CONFIG = METHOD_PREFIX + "LatestConfig";
+
+    private static final ConcurrentHashMap<String, Object> CONFIG_CACHE = new ConcurrentHashMap<>();
+
+    private static final Set<String> LISTENER_KEYS = new HashSet<>();
+
+    private ConcurrentMap<String, HashSet<ConfigurationChangeListener>> configListenersMap =
+        new ConcurrentHashMap<>();
+
+    public static void addConfigListener(String dataId, ConfigurationChangeListener... listeners) {
+        if (StringUtils.isBlank(dataId)) {
+            return;
+        }
+        synchronized (ConfigurationCache.class) {
+            if (!LISTENER_KEYS.contains(dataId)) {
+                ConfigurationFactory.getInstance().addConfigListener(dataId, getInstance());
+                LISTENER_KEYS.add(dataId);
+            }
+            if (null != listeners && listeners.length > 0) {
+                HashSet<ConfigurationChangeListener> listenerHashSet = null;
+                try {
+                    listenerHashSet = getInstance().configListenersMap.get(dataId);
+                    if (CollectionUtils.isEmpty(listenerHashSet)) {
+                        listenerHashSet = new HashSet<>();
+                    }
+                    for (int i = 0; i < listeners.length; i++) {
```

Comment:
For code elegance may use foreach would be better.

Reply:
ok